// @airtable/blocks/dist/cjs/types/mutations.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.MutationTypes = void 0;
/** @module @airtable/blocks: mutations */ /** */

/** @hidden */
var MutationTypes = exports.MutationTypes = Object.freeze({
  SET_MULTIPLE_RECORDS_CELL_VALUES: 'setMultipleRecordsCellValues',
  DELETE_MULTIPLE_RECORDS: 'deleteMultipleRecords',
  CREATE_MULTIPLE_RECORDS: 'createMultipleRecords',
  SET_MULTIPLE_GLOBAL_CONFIG_PATHS: 'setMultipleGlobalConfigPaths',
  CREATE_SINGLE_FIELD: 'createSingleField',
  UPDATE_SINGLE_FIELD_CONFIG: 'updateSingleFieldConfig',
  UPDATE_SINGLE_FIELD_DESCRIPTION: 'updateSingleFieldDescription',
  UPDATE_SINGLE_FIELD_NAME: 'updateSingleFieldName',
  CREATE_SINGLE_TABLE: 'createSingleTable',
  UPDATE_VIEW_METADATA: 'updateViewMetadata'
});

/** @hidden */

/**
 * The Mutation emitted when the App modifies one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/SetMultipleRecordsCellValuesMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App deletes one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/DeleteMultipleRecordsMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App creates one or more {@link Record|Records}.
 *
 * @docsPath testing/mutations/CreateMultipleRecordsMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies one or more values in the
 * {@link GlobalConfig}.
 *
 * @docsPath testing/mutations/SetMultipleGlobalConfigPathsMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App creates a {@link Field}.
 *
 * @docsPath testing/mutations/CreateSingleFieldMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies the options of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldConfigMutation
 */
// note: while this accepts both type and options in config, we enforce liveapp-side that type
// is not changed.

/** @hidden */

/**
 * The Mutation emitted when the App modifies the description of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldDescriptionMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies the name of a {@link Field}.
 *
 * @docsPath testing/mutations/UpdateSingleFieldNameMutation
 */

/** @hidden */

/**
 * Options that affect the behavior of an `UpdateSingleFieldConfigMutation`
 */

/**
 * The Mutation emitted when the App creates a {@link Table}
 *
 * @docsPath testing/mutations/CreateSingleTableMutation
 */

/** @hidden */

/**
 * The Mutation emitted when the App modifies a {@link View}.
 *
 * @docsPath testing/mutations/UpdateViewMetadataMutation
 */

/** @hidden */

/** @hidden */

/** @hidden */

/** */

/** */

/** Indicates whether the user has permission to perform a particular action, and if not, why. */
