// @airtable/blocks/dist/cjs/ui/system/appearance/border.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.config = exports.borderPropTypes = exports.border = void 0;
var _core = require("@styled-system/core");
var _create_style_prop_types = _interopRequireDefault(require("../utils/create_style_prop_types"));
/** @module @airtable/blocks/ui/system: Appearance */ /** */

/** */

var config = exports.config = {
  border: {
    property: 'border',
    scale: 'borders'
  },
  borderWidth: {
    property: 'borderWidth',
    scale: 'borderWidths'
  },
  borderStyle: {
    property: 'borderStyle',
    scale: 'borderStyles'
  },
  borderColor: {
    property: 'borderColor',
    scale: 'colors'
  },
  borderTop: {
    property: 'borderTop',
    scale: 'borders'
  },
  borderRight: {
    property: 'borderRight',
    scale: 'borders'
  },
  borderBottom: {
    property: 'borderBottom',
    scale: 'borders'
  },
  borderLeft: {
    property: 'borderLeft',
    scale: 'borders'
  },
  borderX: {
    properties: ['borderLeft', 'borderRight'],
    scale: 'borders'
  },
  borderY: {
    properties: ['borderTop', 'borderBottom'],
    scale: 'borders'
  }
};
var border = exports.border = (0, _core.system)(config);
var borderPropTypes = exports.borderPropTypes = (0, _create_style_prop_types.default)(border.propNames);
