// @airtable/blocks/dist/cjs/ui/theme/default_theme/switch_variants.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/web.dom-collections.for-each.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _emotion = require("emotion");
var _css_helpers = _interopRequireDefault(require("../../css_helpers"));
var _tokens = require("./tokens");
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
var SWITCH_WIDTH = 20;
var SWITCH_HEIGHT = 12;
var SWITCH_PADDING = 2;
var CIRCLE_SIZE = SWITCH_HEIGHT - 2 * SWITCH_PADDING;
var switchClassName = (0, _emotion.css)({
  width: CIRCLE_SIZE,
  height: CIRCLE_SIZE,
  transition: '0.085s all ease-in',
  borderRadius: _tokens.radii.circle,
  backgroundColor: _tokens.colors.white
});
var switchContainerClassName = (0, _emotion.css)({
  flex: 'none',
  width: SWITCH_WIDTH,
  height: SWITCH_HEIGHT,
  padding: SWITCH_PADDING,
  transition: '0.085s all ease-in',
  boxSizing: 'border-box',
  borderRadius: _tokens.radii.circle,
  backgroundColor: _tokens.colors.darken2
});
var switchLabelClassName = (0, _emotion.css)({
  cursor: 'inherit',
  userSelect: 'none',
  marginLeft: _tokens.space[2],
  color: _tokens.colors.dark,
  flex: 'auto'
});
var baseStyles = (0, _emotion.css)({
  alignItems: 'center',
  borderRadius: _tokens.radii.default,
  boxSizing: 'border-box',
  outline: 'none',
  fontFamily: _tokens.fontFamilies.default,
  backgroundColor: _tokens.colors.lightGray2,
  '&[aria-disabled="true"]': {
    opacity: _tokens.opacities.quieter
  },
  '&:not([aria-disabled="true"])': {
    cursor: 'pointer',
    '&:hover': {
      opacity: _tokens.opacities.quiet
    },
    '&:focus': {
      boxShadow: "inset 0 0 0 2px ".concat(_tokens.colors.darken3)
    }
  },
  ["&[aria-checked=\"true\"] .".concat(switchClassName)]: {
    transform: 'translateX(100%)'
  }
});
var subcomponentClassNames = {
  switchClassName,
  switchContainerClassName,
  switchLabelClassName: (0, _emotion.cx)(switchLabelClassName, _css_helpers.default.TRUNCATE)
};
var switchVariants = {
  default: _objectSpread({
    baseClassName: (0, _emotion.cx)(baseStyles, (0, _emotion.css)({
      ["&[aria-checked=\"true\"] > .".concat(switchContainerClassName)]: {
        backgroundColor: _tokens.colors.greenBright
      }
    }))
  }, subcomponentClassNames),
  danger: _objectSpread({
    baseClassName: (0, _emotion.cx)(baseStyles, (0, _emotion.css)({
      ["&[aria-checked=\"true\"] > .".concat(switchContainerClassName)]: {
        backgroundColor: _tokens.colors.redBright
      }
    }))
  }, subcomponentClassNames)
};
var _default = exports.default = switchVariants;
