// @airtable/blocks/dist/cjs/ui/box.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _index = require("./system/index");
var _aria_props = require("./types/aria_props");
var _tooltip_anchor_props = require("./types/tooltip_anchor_props");
var _data_attributes_prop = require("./types/data_attributes_prop");
var _excluded = ["as", "id", "children", "className", "style", "tabIndex", "role", "onClick", "onMouseEnter", "onMouseLeave", "hasOnClick", "dataAttributes", "aria-label", "aria-labelledby", "aria-describedby", "aria-controls", "aria-expanded", "aria-haspopup", "aria-hidden", "aria-live"];
/** @module @airtable/blocks/ui: Box */
/** */
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Props for the Box component. Also accepts:
 * * {@link AllStylesProps}
 * * {@link AriaProps}
 *
 * @docsPath UI/components/Box
 * @noInheritDoc
 */

/**
 * A box component for creating layouts.
 *
 * [[ Story id="box--example" title="Box example" ]]
 *
 * @component
 * @docsPath UI/components/Box
 */
var Box = (props, ref) => {
  var _props$as = props.as,
    Component = _props$as === void 0 ? 'div' : _props$as,
    id = props.id,
    children = props.children,
    className = props.className,
    style = props.style,
    tabIndex = props.tabIndex,
    role = props.role,
    onClick = props.onClick,
    onMouseEnter = props.onMouseEnter,
    onMouseLeave = props.onMouseLeave,
    hasOnClick = props.hasOnClick,
    _props$dataAttributes = props.dataAttributes,
    dataAttributes = _props$dataAttributes === void 0 ? {} : _props$dataAttributes,
    ariaLabel = props['aria-label'],
    ariaLabelledBy = props['aria-labelledby'],
    ariaDescribedBy = props['aria-describedby'],
    ariaControls = props['aria-controls'],
    ariaExpanded = props['aria-expanded'],
    ariaHasPopup = props['aria-haspopup'],
    ariaHidden = props['aria-hidden'],
    ariaLive = props['aria-live'],
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForStyleProps = (0, _use_styled_system.default)(styleProps);
  return /*#__PURE__*/React.createElement(Component, (0, _extends2.default)({
    ref: ref,
    id: id,
    className: (0, _emotion.cx)(classNameForStyleProps, className),
    style: style,
    tabIndex: tabIndex,
    role: role
    // TODO (stephen): remove tooltip anchor props
    ,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-controls": ariaControls,
    "aria-expanded": ariaExpanded,
    "aria-haspopup": ariaHasPopup,
    "aria-hidden": ariaHidden,
    "aria-live": ariaLive
  }, dataAttributes), children);
};
var ForwardedRefBox = /*#__PURE__*/React.forwardRef(Box);
ForwardedRefBox.propTypes = _objectSpread(_objectSpread(_objectSpread({
  as: _propTypes.default.oneOf(['div', 'span', 'section', 'main', 'nav', 'header', 'footer', 'aside', 'article', 'address', 'hgroup', 'blockquote', 'figure', 'figcaption', 'ol', 'ul', 'li', 'pre']),
  id: _propTypes.default.string,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.number,
  role: _propTypes.default.string,
  dataAttributes: _data_attributes_prop.dataAttributesPropType
}, _aria_props.ariaPropTypes), _tooltip_anchor_props.tooltipAnchorPropTypes), _index.allStylesPropTypes);
ForwardedRefBox.displayName = 'Box';
var _default = exports.default = ForwardedRefBox;
