// @airtable/blocks/dist/cjs/ui/choice_token.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.choiceTokenStylePropTypes = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _core = require("@styled-system/core");
var _baymax_utils = require("./baymax_utils");
var _box = _interopRequireDefault(require("./box"));
var _text = _interopRequireDefault(require("./text"));
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _use_text_color_for_background_color = _interopRequireDefault(require("./use_text_color_for_background_color"));
var _system = require("./system");
var _tooltip_anchor_props = require("./types/tooltip_anchor_props");
var _excluded = ["choice", "onMouseEnter", "onMouseLeave", "onClick", "hasOnClick", "className", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: ChoiceToken */ /** */
/**
 * Style props for the {@link ChoiceToken} component. Accepts:
 * * {@link FlexItemSetProps}
 * * {@link MarginProps}
 * * {@link PositionSetProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.flexItemSet, _system.positionSet, _system.margin);
var choiceTokenStylePropTypes = exports.choiceTokenStylePropTypes = _objectSpread(_objectSpread(_objectSpread({}, _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.marginPropTypes);
var DEFAULT_CHOICE_COLOR = 'gray';

/** An option from a select field. You should not create these objects from scratch, but should instead grab them from base data. */

/**
 * Props for the {@link ChoiceToken} component. Also accepts:
 * * {@link ChoiceTokenStyleProps}
 *
 * @docsPath UI/components/ChoiceToken
 * @noInheritDoc
 */

/**
 * A component that shows a single choice in a small token, to be displayed inline or in a list of choices.
 *
 * [[ Story id="choicetoken--example" title="Choice token example" ]]
 *
 * @component
 * @docsPath UI/components/ChoiceToken
 */
var ChoiceToken = props => {
  var choice = props.choice,
    onMouseEnter = props.onMouseEnter,
    onMouseLeave = props.onMouseLeave,
    onClick = props.onClick,
    hasOnClick = props.hasOnClick,
    className = props.className,
    style = props.style,
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForStyleProps = (0, _use_styled_system.default)(styleProps, styleParser);
  var color = choice.color || DEFAULT_CHOICE_COLOR;
  var textColor = (0, _use_text_color_for_background_color.default)(color);
  return /*#__PURE__*/React.createElement(_box.default, {
    className: (0, _emotion.cx)(className, classNameForStyleProps),
    style: style,
    display: "inline-block"
  }, /*#__PURE__*/React.createElement(_box.default
  // TODO (stephen): remove tooltip anchor props
  , {
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    className: (0, _baymax_utils.baymax)('print-color-exact align-top border-box'),
    backgroundColor: color,
    minWidth: "18px",
    height: "18px",
    borderRadius: "circle",
    paddingX: 2
  }, /*#__PURE__*/React.createElement(_text.default, {
    className: (0, _baymax_utils.baymax)('truncate'),
    textColor: textColor,
    fontSize: "13px",
    fontWeight: "400",
    lineHeight: 1.5
  }, choice.name)));
};
ChoiceToken.propTypes = _objectSpread(_objectSpread({
  choice: _propTypes.default.shape({
    id: _propTypes.default.string,
    name: _propTypes.default.string.isRequired,
    color: _propTypes.default.string
  }).isRequired,
  style: _propTypes.default.object,
  className: _propTypes.default.string
}, _tooltip_anchor_props.tooltipAnchorPropTypes), choiceTokenStylePropTypes);
var _default = exports.default = ChoiceToken;
