// @airtable/blocks/dist/cjs/ui/dialog.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.dialogStylePropTypes = exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _baymax_utils = require("./baymax_utils");
var _modal = _interopRequireDefault(require("./modal"));
var _dialog_close_button = _interopRequireDefault(require("./dialog_close_button"));
var _system = require("./system");
var _excluded = ["onClose", "className", "style", "backgroundClassName", "backgroundStyle", "children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: Dialog */ /** */
/**
 * Style props shared between the {@link Dialog} and {@link ConfirmationDialog} components. Also accepts:
 * * {@link DimensionsSetProps}
 * * {@link FlexContainerSetProps}
 * * {@link SpacingSetProps}
 *
 * @noInheritDoc
 */
// This is equivalent to ModalStyleProps, but we include it separately since Modal is private.

var dialogStylePropTypes = exports.dialogStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.dimensionsSetPropTypes), _system.displayPropTypes), _system.flexContainerSetPropTypes), _system.spacingSetPropTypes);

/**
 * Props for the {@link Dialog} component. Also accepts:
 * * {@link DialogStyleProps}
 *
 * @noInheritDoc
 * @docsPath UI/components/Dialog
 */
/**
 * A styled modal dialog component.
 *
 * [[ Story id="dialog--example" title="Dialog example" ]]
 *
 * @docsPath UI/components/Dialog
 * @component
 */
// TODO (stephen): focus trapping
var Dialog = /*#__PURE__*/function (_React$Component) {
  /** @hidden */
  function Dialog(props) {
    var _this;
    (0, _classCallCheck2.default)(this, Dialog);
    _this = _callSuper(this, Dialog, [props]);
    _this._onKeyDown = _this._onKeyDown.bind(_this);
    return _this;
  }
  /** @hidden */
  (0, _inherits2.default)(Dialog, _React$Component);
  return (0, _createClass2.default)(Dialog, [{
    key: "getChildContext",
    value: function getChildContext() {
      return {
        onDialogClose: this.props.onClose
      };
    }
    /** @hidden */
  }, {
    key: "componentDidMount",
    value: function componentDidMount() {
      window.addEventListener('keydown', this._onKeyDown, false);
    }
    /** @hidden */
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      window.removeEventListener('keydown', this._onKeyDown, false);
    }
    /** @internal */
  }, {
    key: "_onKeyDown",
    value: function _onKeyDown(e) {
      if (e.key === 'Escape') {
        this.props.onClose();
      }
    }
    /** @hidden */
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        onClose = _this$props.onClose,
        className = _this$props.className,
        style = _this$props.style,
        backgroundClassName = _this$props.backgroundClassName,
        backgroundStyle = _this$props.backgroundStyle,
        children = _this$props.children,
        restOfProps = (0, _objectWithoutProperties2.default)(_this$props, _excluded);
      return /*#__PURE__*/React.createElement(_modal.default, (0, _extends2.default)({
        onClose: onClose,
        className: (0, _emotion.cx)((0, _baymax_utils.baymax)('big line-height-4'), className),
        style: style,
        backgroundClassName: backgroundClassName,
        backgroundStyle: backgroundStyle
      }, restOfProps), children);
    }
  }]);
}(React.Component);
/** @hidden */
(0, _defineProperty2.default)(Dialog, "CloseButton", _dialog_close_button.default);
/** @hidden */
(0, _defineProperty2.default)(Dialog, "propTypes", _objectSpread({
  onClose: _propTypes.default.func.isRequired,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  backgroundClassName: _propTypes.default.string,
  backgroundStyle: _propTypes.default.object,
  children: _propTypes.default.node.isRequired
}, dialogStylePropTypes));
// automatically pass onClose to any descendants that are Dialog.CloseButton
/** @hidden */
(0, _defineProperty2.default)(Dialog, "childContextTypes", {
  onDialogClose: _propTypes.default.func
});
var _default = exports.default = Dialog;
