// @airtable/blocks/dist/cjs/ui/dialog_close_button.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DialogCloseButton = void 0;
require("core-js/modules/es.array.includes.js");
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _core = require("@styled-system/core");
var _baymax_utils = require("./baymax_utils");
var _with_styled_system = _interopRequireDefault(require("./with_styled_system"));
var _system = require("./system");
var _tooltip_anchor_props = require("./types/tooltip_anchor_props");
var _icon = _interopRequireDefault(require("./icon"));
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: Dialog */ /** */
/**
 * Props for the {@link DialogCloseButton} component. Also accepts:
 * * {@link DialogCloseButtonStyleProps}
 *
 * @noInheritDoc
 */

/**
 * Style props for the {@link DialogCloseButton} component. Accepts:
 * * {@link BorderRadiusProps}
 * * {@link DimensionsSetProps}
 * * {@link DisplayProps}
 * * {@link FlexContainerSetProps}
 * * {@link FlexItemSetProps}
 * * {@link PositionSetProps}
 * * {@link SpacingSetProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.borderRadius, _system.dimensionsSet, _system.display, _system.flexContainerSet, _system.flexItemSet, _system.positionSet, _system.spacingSet);
var dialogCloseButtonStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.borderRadiusPropTypes), _system.dimensionsSetPropTypes), _system.displayPropTypes), _system.flexContainerSetPropTypes), _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.spacingSetPropTypes);

/**
 * A button that closes {@link Dialog}. Accessed via `Dialog.CloseButton`.
 */
// TODO (stephen): can we alias this as Dialog.CloseButton in typedoc?
// TODO (stephen): convert into a prop in Dialog
var DialogCloseButton = exports.DialogCloseButton = /*#__PURE__*/function (_React$Component) {
  /** @hidden */
  function DialogCloseButton(props) {
    var _this;
    (0, _classCallCheck2.default)(this, DialogCloseButton);
    _this = _callSuper(this, DialogCloseButton, [props]);
    _this._onClick = _this._onClick.bind(_this);
    _this._onKeyDown = _this._onKeyDown.bind(_this);
    return _this;
  }
  /** @internal */
  (0, _inherits2.default)(DialogCloseButton, _React$Component);
  return (0, _createClass2.default)(DialogCloseButton, [{
    key: "_onClick",
    value: function _onClick(e) {
      // onClick should only be defined in the case of a tooltip.
      if (this.props.onClick) {
        this.props.onClick(e);
      }
      this.context.onDialogClose();
    }
    /** @internal */
  }, {
    key: "_onKeyDown",
    value: function _onKeyDown(e) {
      if (e.ctrlKey || e.altKey || e.metaKey) {
        return;
      }
      if (['Enter', ' '].includes(e.key)) {
        e.preventDefault();
        this.context.onDialogClose();
      }
    }
    /** @hidden */
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        onMouseEnter = _this$props.onMouseEnter,
        onMouseLeave = _this$props.onMouseLeave,
        className = _this$props.className,
        style = _this$props.style,
        tabIndex = _this$props.tabIndex,
        children = _this$props.children;
      return /*#__PURE__*/React.createElement("div", {
        // TODO (stephen): remove tooltip anchor props
        onMouseEnter: onMouseEnter,
        onMouseLeave: onMouseLeave,
        onClick: this._onClick,
        onKeyDown: this._onKeyDown,
        className: (0, _emotion.cx)((0, _baymax_utils.baymax)('darken1-hover darken1-focus no-outline pointer'), className),
        style: style,
        tabIndex: tabIndex || 0,
        role: "button",
        "aria-label": "Close dialog"
      }, children ? children : /*#__PURE__*/React.createElement(_icon.default, {
        name: "x",
        size: 12,
        className: (0, _baymax_utils.baymax)('quieter')
      }));
    }
  }]);
}(React.Component);
/** @hidden */
(0, _defineProperty2.default)(DialogCloseButton, "propTypes", _objectSpread({
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  tabIndex: _propTypes.default.number,
  children: _propTypes.default.node
}, _tooltip_anchor_props.tooltipAnchorPropTypes));
/** @hidden */
(0, _defineProperty2.default)(DialogCloseButton, "contextTypes", {
  onDialogClose: _propTypes.default.func
});
var _default = exports.default = (0, _with_styled_system.default)(DialogCloseButton, styleParser, dialogCloseButtonStylePropTypes, {
  position: 'absolute',
  top: 0,
  right: 0,
  marginTop: 2,
  marginRight: 2,
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'center',
  width: '24px',
  height: '24px',
  borderRadius: 'circle'
});
