// @airtable/blocks/dist/cjs/ui/form_field.js
"use strict";

require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.formFieldStylePropTypes = exports.default = void 0;
require("core-js/modules/es.symbol.js");
require("core-js/modules/es.symbol.description.js");
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _react = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var _core = require("@styled-system/core");
var _private_utils = require("../private_utils");
var _box = _interopRequireDefault(require("./box"));
var _text = _interopRequireWildcard(require("./text"));
var _label = _interopRequireDefault(require("./label"));
var _use_form_field = require("./use_form_field");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _system = require("./system");
var _excluded = ["id", "className", "style", "label", "htmlFor", "description", "children"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: FormField */ /** */
/**
 * Style props for the {@link FormField} component. Accepts:
 * * {@link FlexItemSetProps}
 * * {@link MaxWidthProps}
 * * {@link MinWidthProps}
 * * {@link PositionSetProps}
 * * {@link SpacingSetProps}
 * * {@link WidthProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.maxWidth, _system.minWidth, _system.width, _system.flexItemSet, _system.positionSet, _system.spacingSet);
var formFieldStylePropTypes = exports.formFieldStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.maxWidthPropTypes), _system.minWidthPropTypes), _system.widthPropTypes), _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.spacingSetPropTypes);

/**
 * Props for the {@link FormField} component. Also accepts:
 * * {@link FormFieldStyleProps}
 *
 * @noInheritDoc
 * @docsPath UI/components/FormField
 */

/**
 * A form field component that wraps any control field, supplying a provided label and optional
 * description.
 *
 * [[ Story id="formfield--example" title="Form field example" ]]
 *
 * This component will automatically set up the `for` attribute on the outputted label with the `id` attribute
 * on the wrapped control field for the following UI components: Label, Select, FieldPicker,
 * ModelPicker, and ViewPicker. If you'd like to manually override this behavior, you can provide an
 * `htmlFor` prop to this component and manually set the `id` attribute on your wrapped control to
 * the same value.
 *
 * @docsPath UI/components/FormField
 * @component
 */
var FormField = (props, ref) => {
  var id = props.id,
    className = props.className,
    style = props.style,
    _props$label = props.label,
    label = _props$label === void 0 ? _text.TextSize.default : _props$label,
    htmlFor = props.htmlFor,
    description = props.description,
    children = props.children,
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForStyleProps = (0, _use_styled_system.default)(_objectSpread({
    width: '100%'
  }, styleProps), styleParser);
  var _useState = (0, _react.useState)((0, _private_utils.getLocallyUniqueId)('form-field-')),
    _useState2 = (0, _slicedToArray2.default)(_useState, 1),
    generatedId = _useState2[0];
  var controlId = htmlFor || generatedId;
  var _useState3 = (0, _react.useState)((0, _private_utils.getLocallyUniqueId)('input-description-')),
    _useState4 = (0, _slicedToArray2.default)(_useState3, 1),
    generatedDescriptionId = _useState4[0];
  var descriptionId = description ? generatedDescriptionId : '';
  var optionalLabelProps;
  if (description) {
    optionalLabelProps = {
      marginBottom: 1
    };
  }
  return /*#__PURE__*/_react.default.createElement(_box.default, {
    ref: ref,
    id: id,
    className: (0, _emotion.cx)(classNameForStyleProps, className),
    style: style,
    display: "flex",
    flexDirection: "column",
    marginBottom: 3
  }, /*#__PURE__*/_react.default.createElement(_label.default, (0, _extends2.default)({
    htmlFor: controlId,
    size: "default"
  }, optionalLabelProps), label), description && /*#__PURE__*/_react.default.createElement(_text.default, {
    variant: "paragraph",
    id: descriptionId,
    size: "default",
    textColor: "light",
    marginBottom: "6px"
  }, description), /*#__PURE__*/_react.default.createElement(_use_form_field.FormFieldContext.Provider, {
    value: {
      controlId,
      descriptionId
    }
  }, children));
};
var ForwardedRefFormField = /*#__PURE__*/_react.default.forwardRef(FormField);
ForwardedRefFormField.propTypes = _objectSpread({
  id: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  label: _propTypes.default.node,
  htmlFor: _propTypes.default.string,
  description: _propTypes.default.node,
  children: _propTypes.default.node
}, formFieldStylePropTypes);
ForwardedRefFormField.displayName = 'FormField';
var _default = exports.default = ForwardedRefFormField;
