// @airtable/blocks/dist/cjs/ui/heading.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.symbol.description.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.from.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.array.slice.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.regexp.exec.js");
require("core-js/modules/es.regexp.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
require("core-js/modules/es.array.concat.js");
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var _error_utils = require("../error_utils");
var _private_utils = require("../private_utils");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _index = require("./system/index");
var _get_style_props_for_responsive_prop = _interopRequireDefault(require("./system/utils/get_style_props_for_responsive_prop"));
var _use_theme = _interopRequireDefault(require("./theme/use_theme"));
var _aria_props = require("./types/aria_props");
var _data_attributes_prop = require("./types/data_attributes_prop");
var _excluded = ["as", "size", "variant", "children", "id", "role", "dataAttributes", "className", "style", "aria-label", "aria-labelledby", "aria-describedby", "aria-controls", "aria-expanded", "aria-haspopup", "aria-hidden", "aria-live"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
function _createForOfIteratorHelper(r, e) { var t = "undefined" != typeof Symbol && r[Symbol.iterator] || r["@@iterator"]; if (!t) { if (Array.isArray(r) || (t = _unsupportedIterableToArray(r)) || e && r && "number" == typeof r.length) { t && (r = t); var _n = 0, F = function F() {}; return { s: F, n: function n() { return _n >= r.length ? { done: !0 } : { done: !1, value: r[_n++] }; }, e: function e(r) { throw r; }, f: F }; } throw new TypeError("Invalid attempt to iterate non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); } var o, a = !0, u = !1; return { s: function s() { t = t.call(r); }, n: function n() { var r = t.next(); return a = r.done, r; }, e: function e(r) { u = !0, o = r; }, f: function f() { try { a || null == t.return || t.return(); } finally { if (u) throw o; } } }; }
function _unsupportedIterableToArray(r, a) { if (r) { if ("string" == typeof r) return _arrayLikeToArray(r, a); var t = {}.toString.call(r).slice(8, -1); return "Object" === t && r.constructor && (t = r.constructor.name), "Map" === t || "Set" === t ? Array.from(r) : "Arguments" === t || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(t) ? _arrayLikeToArray(r, a) : void 0; } }
function _arrayLikeToArray(r, a) { (null == a || a > r.length) && (a = r.length); for (var e = 0, n = Array(a); e < a; e++) n[e] = r[e]; return n; } /** @module @airtable/blocks/ui: Heading */ /** */
/**
 * Sizes for the {@link Heading} component.
 */

var HeadingSize = (0, _private_utils.createEnum)('xsmall', 'small', 'default', 'large', 'xlarge', 'xxlarge');

/**
 * Size prop for the {@link Heading} component.
 */

var headingSizePropType = (0, _private_utils.createResponsivePropTypeFromEnum)(HeadingSize);

/**
 * Variant prop for the {@link Heading} component.
 * • **default** - Headings typically used for titles.
 * • **caps** - All-caps headings typically used for field names and smaller section headings.
 */

var HeadingVariant = (0, _private_utils.createEnum)('default', 'caps');
var headingVariantPropType = (0, _private_utils.createPropTypeFromEnum)(HeadingVariant);

/** @internal */
function warnIfHeadingSizeOutOfRangeForVariant(size, variant, headingSizesForVariant) {
  if (process.env.NODE_ENV === 'development' && !(0, _private_utils.has)(headingSizesForVariant, size)) {
    // eslint-disable-next-line
    console.warn("[@airtable/blocks/ui] warning: <Heading variant='".concat(variant, "' /> only supports the following values for the size prop: ").concat(Object.keys(headingSizesForVariant).join(', '), ". The component will fall back to the default size."));
  }
}

/** @internal */
function useHeadingStyle(headingSizeProp, variant) {
  var _useTheme = (0, _use_theme.default)(),
    headingStyles = _useTheme.headingStyles;
  var headingSizesForVariant = headingStyles[variant];
  var styleProps;
  if (typeof headingSizeProp === 'string') {
    warnIfHeadingSizeOutOfRangeForVariant(headingSizeProp, variant, headingSizesForVariant);
    styleProps = headingSizesForVariant[headingSizeProp] || headingSizesForVariant[HeadingSize.default];
  } else {
    var responsiveSizePropObject = {};
    var _iterator = _createForOfIteratorHelper((0, _private_utils.keys)(headingSizeProp)),
      _step;
    try {
      for (_iterator.s(); !(_step = _iterator.n()).done;) {
        var sizeKey = _step.value;
        var sizeProp = headingSizeProp[sizeKey];
        (0, _error_utils.invariant)(sizeProp, 'sizeProp');
        warnIfHeadingSizeOutOfRangeForVariant(sizeProp, variant, headingSizesForVariant);
        responsiveSizePropObject[sizeKey] = sizeProp || HeadingSize.default;
      }
    } catch (err) {
      _iterator.e(err);
    } finally {
      _iterator.f();
    }
    styleProps = (0, _get_style_props_for_responsive_prop.default)(responsiveSizePropObject,
    // Cast to any to because `caps` don't all have the same sizes as `default`.
    // This in turn requires each size to be typed as optional.
    headingSizesForVariant);
  }
  return (0, _use_styled_system.default)(styleProps);
}

/**
 * Props for the {@link Heading} component. Also supports:
 * * {@link AllStylesProps}
 * * {@link AriaProps}
 *
 * @docsPath UI/components/Heading
 * @noInheritDoc
 */

/**
 * A heading component with sizes and variants.
 *
 * [[ Story id="heading--example" title="Heading example" ]]
 *
 * @docsPath UI/components/Heading
 * @component
 */
var Heading = (props, ref) => {
  var _props$as = props.as,
    Component = _props$as === void 0 ? 'h3' : _props$as,
    _props$size = props.size,
    size = _props$size === void 0 ? HeadingSize.default : _props$size,
    _props$variant = props.variant,
    variant = _props$variant === void 0 ? HeadingVariant.default : _props$variant,
    children = props.children,
    id = props.id,
    role = props.role,
    dataAttributes = props.dataAttributes,
    className = props.className,
    style = props.style,
    ariaLabel = props['aria-label'],
    ariaLabelledBy = props['aria-labelledby'],
    ariaDescribedBy = props['aria-describedby'],
    ariaControls = props['aria-controls'],
    ariaExpanded = props['aria-expanded'],
    ariaHasPopup = props['aria-haspopup'],
    ariaHidden = props['aria-hidden'],
    ariaLive = props['aria-live'],
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForHeadingSize = useHeadingStyle(size, variant);
  var classNameForStyleProps = (0, _use_styled_system.default)(_objectSpread({
    fontFamily: 'default',
    textColor: 'default'
  }, styleProps));
  return /*#__PURE__*/React.createElement(Component, (0, _extends2.default)({
    ref: ref,
    id: id,
    className: (0, _emotion.cx)(classNameForHeadingSize, classNameForStyleProps, className),
    style: style,
    role: role,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-controls": ariaControls,
    "aria-expanded": ariaExpanded,
    "aria-haspopup": ariaHasPopup,
    "aria-hidden": ariaHidden,
    "aria-live": ariaLive
  }, dataAttributes), children);
};
var ForwardedRefHeading = /*#__PURE__*/React.forwardRef(Heading);
ForwardedRefHeading.propTypes = _objectSpread(_objectSpread({
  as: _propTypes.default.oneOf(['h1', 'h2', 'h3', 'h4', 'h5', 'h6']),
  size: headingSizePropType,
  variant: headingVariantPropType,
  children: _propTypes.default.node,
  id: _propTypes.default.string,
  role: _propTypes.default.string,
  dataAttributes: _data_attributes_prop.dataAttributesPropType,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, _index.allStylesPropTypes), _aria_props.ariaPropTypes);
ForwardedRefHeading.displayName = 'Heading';
var _default = exports.default = ForwardedRefHeading;
