// @airtable/blocks/dist/cjs/ui/input.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.sharedInputPropTypes = exports.inputStylePropTypes = exports.default = exports.ValidInputType = exports.SupportedInputType = void 0;
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.object.to-string.js");
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _core = require("@styled-system/core");
var _private_utils = require("../private_utils");
var _use_theme = _interopRequireDefault(require("./theme/use_theme"));
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _use_form_field = _interopRequireDefault(require("./use_form_field"));
var _system = require("./system");
var _tooltip_anchor_props = require("./types/tooltip_anchor_props");
var _control_sizes = require("./control_sizes");
var _excluded = ["size", "type", "value", "placeholder", "onMouseEnter", "onMouseLeave", "onClick", "hasOnClick", "onChange", "onBlur", "onFocus", "style", "className", "disabled", "required", "spellCheck", "tabIndex", "id", "name", "autoFocus", "max", "maxLength", "min", "minLength", "step", "pattern", "readOnly", "autoComplete", "aria-labelledby", "aria-describedby"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: Input */ /** */
/** @internal */

var InputVariant = (0, _private_utils.createEnum)('default');

/** @internal */
function useInputVariant() {
  var variant = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : InputVariant.default;
  var _useTheme = (0, _use_theme.default)(),
    inputVariants = _useTheme.inputVariants;
  return inputVariants[variant];
}
var ValidInputType = exports.ValidInputType = (0, _private_utils.createEnum)('date', 'datetime-local', 'email', 'month', 'number', 'password', 'search', 'tel', 'text', 'time', 'url', 'week');
/** */

/**
 * Style props shared between the {@link Input} and {@link InputSynced} components. Accepts:
 * * {@link FlexItemSetProps}
 * * {@link MarginProps}
 * * {@link MaxWidthProps}
 * * {@link MinWidthProps}
 * * {@link PositionSetProps}
 * * {@link WidthProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.maxWidth, _system.minWidth, _system.width, _system.flexItemSet, _system.positionSet, _system.margin);
var inputStylePropTypes = exports.inputStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.maxWidthPropTypes), _system.minWidthPropTypes), _system.widthPropTypes), _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.marginPropTypes);

/**
 * Props shared between the {@link Input} and {@link InputSynced} components.
 *
 * @noInheritDoc
 */

var SupportedInputType = exports.SupportedInputType = (0, _private_utils.createEnum)('date', 'datetime-local', 'email', 'month', 'number', 'password', 'search', 'tel', 'text', 'time', 'url', 'week');
/**
 * Supported types for the {@link Input} component. See [MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input#%3Cinput%3E_types) for more information.
 */

// Shared with `Input` and `InputSynced`.
var sharedInputPropTypes = exports.sharedInputPropTypes = _objectSpread(_objectSpread({
  size: _control_sizes.controlSizePropType,
  type: (0, _private_utils.createPropTypeFromEnum)(SupportedInputType),
  placeholder: _propTypes.default.string,
  disabled: _propTypes.default.bool,
  required: _propTypes.default.bool,
  spellCheck: _propTypes.default.bool,
  tabIndex: _propTypes.default.oneOfType([_propTypes.default.number]),
  autoFocus: _propTypes.default.bool,
  max: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  maxLength: _propTypes.default.number,
  min: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  minLength: _propTypes.default.number,
  step: _propTypes.default.oneOfType([_propTypes.default.number, _propTypes.default.string]),
  pattern: _propTypes.default.string,
  readOnly: _propTypes.default.bool,
  autoComplete: _propTypes.default.string,
  onChange: _propTypes.default.func,
  onBlur: _propTypes.default.func,
  onFocus: _propTypes.default.func,
  style: _propTypes.default.object,
  className: _propTypes.default.string,
  'aria-labelledby': _propTypes.default.string,
  'aria-describedby': _propTypes.default.string
}, inputStylePropTypes), _tooltip_anchor_props.tooltipAnchorPropTypes);

/**
 * Props for the {@link Input} component. Also accepts:
 * * {@link InputStyleProps}
 *
 * @docsPath UI/components/Input
 */

/**
 * An input component. A wrapper around `<input>` that fits in with Airtable's user interface.
 *
 * [[ Story id="input--example" title="Input example" ]]
 *
 * @docsPath UI/components/Input
 * @component
 */
var Input = (props, ref) => {
  var _props$size = props.size,
    size = _props$size === void 0 ? _control_sizes.ControlSize.default : _props$size,
    _props$type = props.type,
    type = _props$type === void 0 ? SupportedInputType.text : _props$type,
    value = props.value,
    placeholder = props.placeholder,
    onMouseEnter = props.onMouseEnter,
    onMouseLeave = props.onMouseLeave,
    onClick = props.onClick,
    hasOnClick = props.hasOnClick,
    onChange = props.onChange,
    onBlur = props.onBlur,
    onFocus = props.onFocus,
    style = props.style,
    className = props.className,
    disabled = props.disabled,
    required = props.required,
    spellCheck = props.spellCheck,
    tabIndex = props.tabIndex,
    id = props.id,
    name = props.name,
    autoFocus = props.autoFocus,
    max = props.max,
    maxLength = props.maxLength,
    min = props.min,
    minLength = props.minLength,
    step = props.step,
    pattern = props.pattern,
    readOnly = props.readOnly,
    autoComplete = props.autoComplete,
    ariaLabelledBy = props['aria-labelledby'],
    ariaDescribedByProp = props['aria-describedby'],
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var formFieldContextValue = (0, _use_form_field.default)();
  var controlId = formFieldContextValue ? formFieldContextValue.controlId : undefined;
  var descriptionId = formFieldContextValue ? formFieldContextValue.descriptionId : undefined;
  var ariaDescribedBy = [ariaDescribedByProp, descriptionId].filter(Boolean).join(' ') || undefined;
  // There is only a single default variant.
  var classNameForInputVariant = useInputVariant();
  var classNameForInputSize = (0, _control_sizes.useInputSize)(size);
  var classNameForStyleProps = (0, _use_styled_system.default)(_objectSpread({
    width: '100%'
  }, styleProps), styleParser);
  return /*#__PURE__*/React.createElement("input", {
    ref: ref,
    value: value,
    type: type,
    placeholder: placeholder,
    disabled: disabled,
    required: required,
    spellCheck: spellCheck,
    tabIndex: tabIndex,
    id: id || controlId,
    name: name,
    autoFocus: autoFocus,
    max: max,
    maxLength: maxLength,
    min: min,
    minLength: minLength,
    step: step,
    pattern: pattern,
    readOnly: readOnly,
    autoComplete: autoComplete,
    onChange: onChange,
    onBlur: onBlur,
    onFocus: onFocus
    // TODO (stephen): remove tooltip anchor props
    ,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    style: style,
    className: (0, _emotion.cx)(classNameForInputSize, classNameForInputVariant, classNameForStyleProps, className),
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy
  });
};
var ForwardedRefInput = /*#__PURE__*/React.forwardRef(Input);
ForwardedRefInput.propTypes = _objectSpread({
  value: _propTypes.default.string.isRequired
}, sharedInputPropTypes);
ForwardedRefInput.displayName = 'Input';
var _default = exports.default = ForwardedRefInput;
