// @airtable/blocks/dist/cjs/ui/label.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _index = require("./system/index");
var _aria_props = require("./types/aria_props");
var _text = require("./text");
var _data_attributes_prop = require("./types/data_attributes_prop");
var _excluded = ["size", "htmlFor", "id", "children", "className", "style", "dataAttributes", "role", "aria-label", "aria-labelledby", "aria-describedby", "aria-controls", "aria-expanded", "aria-haspopup", "aria-hidden", "aria-live"];
/** @module @airtable/blocks/ui: Label */
/** */
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Props for the {@link Label} component. Also accepts:
 * * {@link AllStylesProps}
 * * {@link AriaProps}
 *
 * @noInheritDoc
 * @docsPath UI/components/Label
 */

/**
 * A label component.
 *
 * [[ Story id="label--example" title="Label example" ]]
 *
 * @docsPath UI/components/Label
 * @component
 */
var Label = (props, ref) => {
  var _props$size = props.size,
    size = _props$size === void 0 ? _text.TextSize.default : _props$size,
    htmlFor = props.htmlFor,
    id = props.id,
    children = props.children,
    className = props.className,
    style = props.style,
    dataAttributes = props.dataAttributes,
    role = props.role,
    ariaLabel = props['aria-label'],
    ariaLabelledBy = props['aria-labelledby'],
    ariaDescribedBy = props['aria-describedby'],
    ariaControls = props['aria-controls'],
    ariaExpanded = props['aria-expanded'],
    ariaHasPopup = props['aria-haspopup'],
    ariaHidden = props['aria-hidden'],
    ariaLive = props['aria-live'],
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForTextStyle = (0, _text.useTextStyle)(size);
  var classNameForStyleProps = (0, _use_styled_system.default)(_objectSpread({
    display: 'inline-block',
    textColor: 'light',
    fontWeight: 'strong',
    marginBottom: '6px'
  }, styleProps));
  return /*#__PURE__*/React.createElement("label", (0, _extends2.default)({
    ref: ref,
    htmlFor: htmlFor,
    id: id,
    className: (0, _emotion.cx)(classNameForTextStyle, classNameForStyleProps, className),
    style: style,
    role: role,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-controls": ariaControls,
    "aria-expanded": ariaExpanded,
    "aria-haspopup": ariaHasPopup,
    "aria-hidden": ariaHidden,
    "aria-live": ariaLive
  }, dataAttributes), children);
};
var ForwardedRefLabel = /*#__PURE__*/React.forwardRef(Label);
ForwardedRefLabel.propTypes = _objectSpread(_objectSpread({
  size: _text.textSizePropType,
  htmlFor: _propTypes.default.string,
  id: _propTypes.default.string,
  dataAttributes: _data_attributes_prop.dataAttributesPropType,
  children: _propTypes.default.node,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, _index.allStylesPropTypes), _aria_props.ariaPropTypes);
ForwardedRefLabel.displayName = 'Label';
var _default = exports.default = ForwardedRefLabel;
