// @airtable/blocks/dist/cjs/ui/loader.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loaderStylePropTypes = exports.default = void 0;
require("core-js/modules/es.array.concat.js");
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _core = require("@styled-system/core");
var _baymax_utils = require("./baymax_utils");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _system = require("./system");
var _excluded = ["fillColor", "scale", "className", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: Loader */ /** */
var ORIGINAL_SIZE = 54;

/**
 * Style props for the {@link Loader} component. Accepts:
 * * {@link FlexItemSetProps}
 * * {@link MarginProps}
 * * {@link PositionSetProps}
 */

var styleParser = (0, _core.compose)(_system.flexItemSet, _system.positionSet, _system.margin);
var loaderStylePropTypes = exports.loaderStylePropTypes = _objectSpread(_objectSpread(_objectSpread({}, _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.marginPropTypes);

/**
 * Props for the {@link Loader} component. Also accepts:
 * * {@link LoaderStyleProps}
 *
 * @docsPath UI/components/Loader
 * @noInheritDoc
 */

// Override the default props and then just proxy through to our loader.
/**
 * A loading spinner component.
 *
 * [[ Story id="loader--example" title="Loader example" ]]
 *
 * @docsPath UI/components/Loader
 * @component
 */
var Loader = props => {
  var fillColor = props.fillColor,
    scale = props.scale,
    className = props.className,
    style = props.style,
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForStyleProps = (0, _use_styled_system.default)(styleProps, styleParser);
  return /*#__PURE__*/React.createElement("svg", {
    width: ORIGINAL_SIZE * scale,
    height: ORIGINAL_SIZE * scale,
    viewBox: "0 0 ".concat(ORIGINAL_SIZE, " ").concat(ORIGINAL_SIZE),
    className: (0, _emotion.cx)((0, _baymax_utils.baymax)('animate-spin-scale animate-infinite'), classNameForStyleProps, className),
    style: _objectSpread({
      shapeRendering: 'geometricPrecision'
    }, style)
  }, /*#__PURE__*/React.createElement("g", null, /*#__PURE__*/React.createElement("path", {
    d: "M10.9,48.6c-1.6-1.3-2-3.6-0.7-5.3c1.3-1.6,3.6-2.1,5.3-0.8c0.8,0.5,1.5,1.1,2.4,1.5c7.5,4.1,16.8,2.7,22.8-3.4c1.5-1.5,3.8-1.5,5.3,0c1.4,1.5,1.4,3.9,0,5.3c-8.4,8.5-21.4,10.6-31.8,4.8C13,50.1,11.9,49.3,10.9,48.6z",
    fill: fillColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M53.6,31.4c-0.3,2.1-2.3,3.5-4.4,3.2c-2.1-0.3-3.4-2.3-3.1-4.4c0.2-1.1,0.2-2.2,0.2-3.3c0-8.7-5.7-16.2-13.7-18.5c-2-0.5-3.2-2.7-2.6-4.7s2.6-3.2,4.7-2.6C46,4.4,53.9,14.9,53.9,27C53.9,28.5,53.8,30,53.6,31.4z",
    fill: fillColor
  }), /*#__PURE__*/React.createElement("path", {
    d: "M16.7,1.9c1.9-0.8,4.1,0.2,4.8,2.2s-0.2,4.2-2.1,5c-7.2,2.9-12,10-12,18.1c0,1.6,0.2,3.2,0.6,4.7c0.5,2-0.7,4.1-2.7,4.6c-2,0.5-4-0.7-4.5-2.8C0.3,31.5,0,29.3,0,27.1C0,15.8,6.7,5.9,16.7,1.9z",
    fill: fillColor
  })));
};
Loader.propTypes = _objectSpread({
  fillColor: _propTypes.default.string.isRequired,
  scale: _propTypes.default.number.isRequired,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, loaderStylePropTypes);
Loader.defaultProps = {
  fillColor: '#888',
  scale: 0.3
};
var _default = exports.default = Loader;
