// @airtable/blocks/dist/cjs/ui/modal.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.modalStylePropTypes = exports.default = exports.Modal = void 0;
var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));
var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));
var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));
var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));
var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _reactDom = _interopRequireDefault(require("react-dom"));
var _core = require("@styled-system/core");
var _error_utils = require("../error_utils");
var _baymax_utils = require("./baymax_utils");
var _with_styled_system = _interopRequireDefault(require("./with_styled_system"));
var _system = require("./system");
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function _callSuper(t, o, e) { return o = (0, _getPrototypeOf2.default)(o), (0, _possibleConstructorReturn2.default)(t, _isNativeReflectConstruct() ? Reflect.construct(o, e || [], (0, _getPrototypeOf2.default)(t).constructor) : o.apply(t, e)); }
function _isNativeReflectConstruct() { try { var t = !Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); } catch (t) {} return (_isNativeReflectConstruct = function _isNativeReflectConstruct() { return !!t; })(); }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @hidden */ /** */
/**
 * Props for the {@link Modal} component. Also accepts:
 * * {@link ModalStyleProps}
 *
 * @hidden
 */
// This doesn't actually extend ModalStyleProps since withStyledSystem
// expects non-style props and style props as separate generic type variables.

/**
 * Style props shared between the {@link Modal}, {@link Dialog}, and {@link ConfirmationDialog} components. Also accepts:
 * * {@link DimensionsSetProps}
 * * {@link FlexContainerSetProps}
 * * {@link SpacingSetProps}
 *
 * @hidden
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.dimensionsSet, _system.display, _system.flexContainerSet, _system.spacingSet);
var modalStylePropTypes = exports.modalStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.dimensionsSetPropTypes), _system.displayPropTypes), _system.flexContainerSetPropTypes), _system.spacingSetPropTypes);

/**
 * Generic modal component with minimal styling.
 *
 * @hidden
 */
// TODO (stephen): refactor so Modal only includes the background element and renders the dialog via children
var Modal = exports.Modal = /*#__PURE__*/function (_React$Component) {
  /** @hidden */
  function Modal(props) {
    var _this;
    (0, _classCallCheck2.default)(this, Modal);
    _this = _callSuper(this, Modal, [props]);
    /** @internal */
    (0, _defineProperty2.default)(_this, "_background", void 0);
    /** @internal */
    (0, _defineProperty2.default)(_this, "_container", void 0);
    /** @internal */
    (0, _defineProperty2.default)(_this, "_originalActiveElement", void 0);
    /** @internal */
    (0, _defineProperty2.default)(_this, "_mouseDownOutsideModal", void 0);
    _this._background = null;
    _this._container = document.createElement('div');
    _this._originalActiveElement = null;
    _this._mouseDownOutsideModal = false;
    _this._onMouseDown = _this._onMouseDown.bind(_this);
    _this._onMouseUp = _this._onMouseUp.bind(_this);
    return _this;
  }
  /** @hidden */
  (0, _inherits2.default)(Modal, _React$Component);
  return (0, _createClass2.default)(Modal, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var container = this._container;
      container.setAttribute('tabIndex', '0');
      container.style.zIndex = '99999';
      // If we don't set `position: fixed`, the outline and box-shadow
      // of elements that are in theory underneath this element will cover
      // up the modal.
      container.style.position = 'fixed';
      (0, _error_utils.invariant)(document.body, 'no document body');
      document.body.appendChild(container);

      // If the frame is focused, move focus to the modal's container.
      // Next time the user presses tab, it will focus the first focusable element in the modal.
      // We only do this if the document is focused to avoid the frame becoming
      // programmatically focused if a modal is displayed without user interaction.
      if (document.hasFocus()) {
        this._originalActiveElement = document.activeElement;
        container.focus();
      }
    }
    /** @hidden */
  }, {
    key: "componentWillUnmount",
    value: function componentWillUnmount() {
      (0, _error_utils.invariant)(document.body, 'no document body');
      document.body.removeChild(this._container);
      if (this._originalActiveElement !== null) {
        this._originalActiveElement.focus();
      }
    }
    /** @internal */
  }, {
    key: "_onMouseDown",
    value: function _onMouseDown(e) {
      if (this._shouldClickingOnElementCloseModal(e.target)) {
        this._mouseDownOutsideModal = true;
      }
    }
    /** @internal */
  }, {
    key: "_onMouseUp",
    value: function _onMouseUp(e) {
      if (this._mouseDownOutsideModal && this.props.onClose && this._shouldClickingOnElementCloseModal(e.target)) {
        this.props.onClose();
      }
      this._mouseDownOutsideModal = false;
    }
    /** @internal */
  }, {
    key: "_shouldClickingOnElementCloseModal",
    value: function _shouldClickingOnElementCloseModal(el) {
      return el === this._background;
    }
    /** @hidden */
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
        className = _this$props.className,
        style = _this$props.style,
        backgroundClassName = _this$props.backgroundClassName,
        backgroundStyle = _this$props.backgroundStyle,
        children = _this$props.children;
      return /*#__PURE__*/_reactDom.default.createPortal( /*#__PURE__*/React.createElement("div", {
        ref: el => this._background = el,
        className: (0, _emotion.cx)((0, _baymax_utils.baymax)('fixed all-0 darken3 flex items-center justify-center'), backgroundClassName),
        style: backgroundStyle,
        onMouseDown: this._onMouseDown,
        onMouseUp: this._onMouseUp
      }, /*#__PURE__*/React.createElement("div", {
        className: (0, _emotion.cx)((0, _baymax_utils.baymax)('relative overflow-auto light-scrollbar white stroked1 rounded-big animate-bounce-in'), className),
        style: style
      }, children)), this._container);
    }
  }]);
}(React.Component);
/** @hidden */
(0, _defineProperty2.default)(Modal, "propTypes", {
  onClose: _propTypes.default.func,
  className: _propTypes.default.string,
  style: _propTypes.default.object,
  backgroundClassName: _propTypes.default.string,
  backgroundStyle: _propTypes.default.object
});
var _default = exports.default = (0, _with_styled_system.default)(Modal, styleParser, modalStylePropTypes,
// TODO (stephen): move these to Dialog
{
  display: 'block',
  width: '100%',
  maxWidth: '100vw',
  maxHeight: '100vh',
  margin: 3,
  padding: 3
});
