// @airtable/blocks/dist/cjs/ui/progress_bar.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.progressBarStylePropTypes = exports.default = void 0;
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var React = _interopRequireWildcard(require("react"));
var _core = require("@styled-system/core");
var _colors = _interopRequireDefault(require("../colors"));
var _private_utils = require("../private_utils");
var _baymax_utils = require("./baymax_utils");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _system = require("./system");
var _tooltip_anchor_props = require("./types/tooltip_anchor_props");
var _box = _interopRequireDefault(require("./box"));
var _default_theme = _interopRequireDefault(require("./theme/default_theme"));
var _excluded = ["barColor", "progress", "onMouseEnter", "onMouseLeave", "onClick", "hasOnClick", "className", "style"];
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; } /** @module @airtable/blocks/ui: ProgressBar */ /** */
/**
 * Style props for the {@link ProgressBar} component. Also accepts:
 * * {@link BackgroundColorProps}
 * * {@link FlexItemSetProps}
 * * {@link HeightProps}
 * * {@link MarginProps}
 * * {@link MaxWidthProps}
 * * {@link MinWidthProps}
 * * {@link PositionSetProps}
 * * {@link WidthProps}
 *
 * @noInheritDoc
 */

var styleParser = (0, _core.compose)(_system.backgroundColor, _system.maxWidth, _system.minWidth, _system.width, _system.height, _system.flexItemSet, _system.positionSet, _system.margin, _system.display);
var progressBarStylePropTypes = exports.progressBarStylePropTypes = _objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread(_objectSpread({}, _system.backgroundColorPropTypes), _system.maxWidthPropTypes), _system.minWidthPropTypes), _system.widthPropTypes), _system.heightPropTypes), _system.flexItemSetPropTypes), _system.positionSetPropTypes), _system.marginPropTypes), _system.displayPropTypes);

/**
 * Props for the {@link ProgressBar} component. Also accepts:
 * * {@link ProgressBarStyleProps}
 *
 * @docsPath UI/components/ProgressBar
 * @noInheritDoc
 */

// TODO (jay): We can't use Box yet, because `aria-valuenow` is not yet available on `AriaProps`.
var progressBarClassName = (0, _emotion.css)({
  position: 'relative',
  borderRadius: _default_theme.default.radii.circle,
  overflow: 'hidden'
});

/**
 * A progress bar.
 *
 * @example
 * ```js
 * import {ProgressBar} from '@airtable/blocks/ui';
 *
 * function MyComponent() {
 *     return (
 *         <ProgressBar
 *             progress={0.6}
 *             barColor='#ff9900'
 *         />
 *     );
 * }
 * ```
 * @docsPath UI/components/ProgressBar
 * @component
 */
var ProgressBar = props => {
  var _props$barColor = props.barColor,
    barColor = _props$barColor === void 0 ? _default_theme.default.colors.blueBright : _props$barColor,
    progress = props.progress,
    onMouseEnter = props.onMouseEnter,
    onMouseLeave = props.onMouseLeave,
    onClick = props.onClick,
    hasOnClick = props.hasOnClick,
    className = props.className,
    style = props.style,
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var clampedProgressValue = (0, _private_utils.clamp)(progress, 0, 1) * 100;
  var classNameForStyleProps = (0, _use_styled_system.default)(styleProps, styleParser);
  return /*#__PURE__*/React.createElement("div", {
    role: "progressbar",
    "aria-valuenow": clampedProgressValue
    // TODO (stephen): remove tooltip anchor props
    ,
    onMouseEnter: onMouseEnter,
    onMouseLeave: onMouseLeave,
    onClick: onClick,
    className: (0, _emotion.cx)(progressBarClassName, classNameForStyleProps, className),
    style: style
  }, /*#__PURE__*/React.createElement(_box.default, {
    className: (0, _baymax_utils.baymax)('animate'),
    width: "".concat(clampedProgressValue, "%"),
    height: "100%",
    backgroundColor: barColor
  }));
};
ProgressBar.propTypes = _objectSpread(_objectSpread({
  progress: _propTypes.default.number.isRequired,
  barColor: _propTypes.default.string,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, _tooltip_anchor_props.tooltipAnchorPropTypes), progressBarStylePropTypes);
ProgressBar.defaultProps = {
  barColor: _colors.default.BLUE_BRIGHT,
  backgroundColor: _colors.default.GRAY_LIGHT_1,
  width: '100%',
  height: 4
};
var _default = exports.default = ProgressBar;
