// @airtable/blocks/dist/cjs/ui/text.js
"use strict";

require("core-js/modules/es.symbol.js");
require("core-js/modules/es.array.filter.js");
require("core-js/modules/es.array.iterator.js");
require("core-js/modules/es.object.get-own-property-descriptors.js");
require("core-js/modules/es.object.to-string.js");
require("core-js/modules/es.weak-map.js");
require("core-js/modules/web.dom-collections.for-each.js");
require("core-js/modules/web.dom-collections.iterator.js");
var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");
Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.textVariantPropType = exports.textSizePropType = exports.default = exports.TextVariant = exports.TextSize = void 0;
exports.useTextStyle = useTextStyle;
var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));
var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));
var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));
var React = _interopRequireWildcard(require("react"));
var _propTypes = _interopRequireDefault(require("prop-types"));
var _emotion = require("emotion");
var _private_utils = require("../private_utils");
var _use_styled_system = _interopRequireDefault(require("./use_styled_system"));
var _index = require("./system/index");
var _get_style_props_for_responsive_prop = _interopRequireDefault(require("./system/utils/get_style_props_for_responsive_prop"));
var _use_theme = _interopRequireDefault(require("./theme/use_theme"));
var _aria_props = require("./types/aria_props");
var _data_attributes_prop = require("./types/data_attributes_prop");
var _excluded = ["as", "size", "variant", "children", "id", "role", "dataAttributes", "className", "style", "aria-label", "aria-labelledby", "aria-describedby", "aria-controls", "aria-expanded", "aria-haspopup", "aria-hidden", "aria-live"];
/** @module @airtable/blocks/ui: Text */
/** */
function _getRequireWildcardCache(e) { if ("function" != typeof WeakMap) return null; var r = new WeakMap(), t = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(e) { return e ? t : r; })(e); }
function _interopRequireWildcard(e, r) { if (!r && e && e.__esModule) return e; if (null === e || "object" != typeof e && "function" != typeof e) return { default: e }; var t = _getRequireWildcardCache(r); if (t && t.has(e)) return t.get(e); var n = { __proto__: null }, a = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var u in e) if ("default" !== u && {}.hasOwnProperty.call(e, u)) { var i = a ? Object.getOwnPropertyDescriptor(e, u) : null; i && (i.get || i.set) ? Object.defineProperty(n, u, i) : n[u] = e[u]; } return n.default = e, t && t.set(e, n), n; }
function ownKeys(e, r) { var t = Object.keys(e); if (Object.getOwnPropertySymbols) { var o = Object.getOwnPropertySymbols(e); r && (o = o.filter(function (r) { return Object.getOwnPropertyDescriptor(e, r).enumerable; })), t.push.apply(t, o); } return t; }
function _objectSpread(e) { for (var r = 1; r < arguments.length; r++) { var t = null != arguments[r] ? arguments[r] : {}; r % 2 ? ownKeys(Object(t), !0).forEach(function (r) { (0, _defineProperty2.default)(e, r, t[r]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(e, Object.getOwnPropertyDescriptors(t)) : ownKeys(Object(t)).forEach(function (r) { Object.defineProperty(e, r, Object.getOwnPropertyDescriptor(t, r)); }); } return e; }
/**
 * Variants for the {@link Text} component:
 *
 * • **default**
 *
 * Single-line text.
 *
 * • **paragraph**
 *
 * Multi-line text such as body copy.
 */

var TextVariant = exports.TextVariant = (0, _private_utils.createEnum)('default', 'paragraph');
var textVariantPropType = exports.textVariantPropType = (0, _private_utils.createPropTypeFromEnum)(TextVariant);

/**
 * Sizes for the {@link Text} component.
 */

var TextSize = exports.TextSize = (0, _private_utils.createEnum)('small', 'default', 'large', 'xlarge');
/** */

var textSizePropType = exports.textSizePropType = (0, _private_utils.createResponsivePropTypeFromEnum)(TextSize);

/** @internal */
function useTextStyle(textSizeProp) {
  var variant = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : TextVariant.default;
  var _useTheme = (0, _use_theme.default)(),
    textStyles = _useTheme.textStyles;
  var textSizesForVariant = textStyles[variant];
  var styleProps;
  if (typeof textSizeProp === 'string') {
    styleProps = textSizesForVariant[textSizeProp];
  } else {
    styleProps = (0, _get_style_props_for_responsive_prop.default)(textSizeProp, textSizesForVariant);
  }
  return (0, _use_styled_system.default)(styleProps);
}

/**
 * Props for the {@link Text} component. Also supports:
 * * {@link AriaProps}
 * * {@link AllStylesProps}
 *
 * @noInheritDoc
 * @docsPath UI/components/Text
 */

/**
 * A text component with sizes and variants.
 *
 * [[ Story id="text--example" title="Text example" ]]
 *
 * @docsPath UI/components/Text
 * @component
 */
var Text = (props, ref) => {
  var _props$as = props.as,
    Component = _props$as === void 0 ? 'p' : _props$as,
    _props$size = props.size,
    size = _props$size === void 0 ? TextSize.default : _props$size,
    _props$variant = props.variant,
    variant = _props$variant === void 0 ? TextVariant.default : _props$variant,
    children = props.children,
    id = props.id,
    role = props.role,
    dataAttributes = props.dataAttributes,
    className = props.className,
    style = props.style,
    ariaLabel = props['aria-label'],
    ariaLabelledBy = props['aria-labelledby'],
    ariaDescribedBy = props['aria-describedby'],
    ariaControls = props['aria-controls'],
    ariaExpanded = props['aria-expanded'],
    ariaHasPopup = props['aria-haspopup'],
    ariaHidden = props['aria-hidden'],
    ariaLive = props['aria-live'],
    styleProps = (0, _objectWithoutProperties2.default)(props, _excluded);
  var classNameForTextStyle = useTextStyle(size, variant);
  var classNameForStyleProps = (0, _use_styled_system.default)(_objectSpread({
    textColor: 'default',
    fontFamily: 'default'
  }, styleProps));
  return /*#__PURE__*/React.createElement(Component, (0, _extends2.default)({
    ref: ref,
    id: id,
    className: (0, _emotion.cx)(classNameForTextStyle, classNameForStyleProps, className),
    style: style,
    role: role,
    "aria-label": ariaLabel,
    "aria-labelledby": ariaLabelledBy,
    "aria-describedby": ariaDescribedBy,
    "aria-controls": ariaControls,
    "aria-expanded": ariaExpanded,
    "aria-haspopup": ariaHasPopup,
    "aria-hidden": ariaHidden,
    "aria-live": ariaLive
  }, dataAttributes), children);
};
var ForwardedRefText = /*#__PURE__*/React.forwardRef(Text);
ForwardedRefText.propTypes = _objectSpread(_objectSpread({
  as: _propTypes.default.oneOf(['p', 'h1', 'h2', 'h3', 'h4', 'h5', 'h6', 'span', 'li', 'em', 'strong', 'kbd', 'mark', 'q', 's', 'samp', 'small', 'sub', 'sup', 'time', 'var', 'blockquote']),
  size: textSizePropType,
  variant: textVariantPropType,
  children: _propTypes.default.node,
  id: _propTypes.default.string,
  role: _propTypes.default.string,
  dataAttributes: _data_attributes_prop.dataAttributesPropType,
  className: _propTypes.default.string,
  style: _propTypes.default.object
}, _index.allStylesPropTypes), _aria_props.ariaPropTypes);
ForwardedRefText.displayName = 'Text';
var _default = exports.default = ForwardedRefText;
